Unified FlashPaletteEffect and exposed it to Lua

Mailaender opened the pull request:
Split from https://github.com/OpenRA/OpenRA/pull/7863.

Mailaender committed:
add a Lua controllable flash palette trait

Phrohdoh commented:
Shouldn't this subclass the global modifier?

Mailaender commented:
Why that?

ScottNZ commented on OpenRA.Mods.Common/Traits/PaletteEffects/FlashPaletteEffect.cs:
```diff
@@ -0,0 +1,61 @@
+	{
+		public readonly string[] ExcludePalettes = { "cursor", "chrome", "colorpicker", "fog", "shroud" };
+
+		public readonly float LightningChance = 0.99f;
```
I think this should be `0.01f` (or use the integer percentage stuff and `1`)

Phrohdoh replied:
I agree with this.
A lower chance value naturally implies the even will (probably) occur less often.

Phrohdoh commented:
Because you are just duplicating some logic.
Especially the info property, but perhaps that is not best.

Phrohdoh commented on OpenRA.Mods.Common/Traits/Sound/Thunder.cs:
```diff
@@ -0,0 +1,41 @@
+		public object Create(ActorInitializer init) { return new Thunder(init, this); }
+	}
+
+	class Thunder : ITick
```
`Thunder` is pretty specific when in all actuality you are just playing ambient audio.

Mailaender replied:
This plays an audio whenever the flash palette flashes. See https://github.com/OpenRA/OpenRA/pull/7871 for regular ambient audio.

Phrohdoh replied:
True, ambient was a bad word on my part, but my point still stands.
Being that this isn't necessarily a thunder sound, it is just a sound played on a condition (which is determined by another trait).

Mailaender replied:
Suggest an alternative then.

Phrohdoh commented on OpenRA.Mods.Common/Traits/PaletteEffects/FlashPaletteEffect.cs:
```diff
@@ -0,0 +1,61 @@
+
+				for (var x = 0; x < 256; x++)
+				{
+					if (Lightning)
```
Wouldn't it be more performant to do:

```csharp
if (Lightning)
    for ..
```

Mailaender replied:
I had thoughts that might be necessary for the effect which combined everything in one loop, but it seems the lightning flash still works.

Phrohdoh commented on OpenRA.Mods.Common/Traits/PaletteEffects/FlashPaletteEffect.cs:
```diff
@@ -0,0 +1,61 @@
+
+				var palette = kvp.Value;
+
+				for (var x = 0; x < 256; x++)
```
We have the static `Palette.Size`, do we want to use it or an explicit `256` in these traits?

penev92 replied:
I really don't like magic numbers, however obvious they are, so I say use the static field.

Mailaender replied:
Done.

Phrohdoh commented on OpenRA.Mods.Common/Traits/PaletteEffects/FlashPaletteEffect.cs:
```diff
@@ -0,0 +1,61 @@
+
+	class FlashPaletteEffect : IPaletteModifier, ITick
+	{
+		public bool Lightning = false;
```
The class name is pretty generic, so this should be too IMO.

Mailaender replied:
Okay, why not, but this is not exposed to modders anywhere.

Phrohdoh commented on OpenRA.Mods.Common/Traits/Sound/Thunder.cs:
```diff
@@ -0,0 +1,41 @@
+
+		public void Tick(Actor self)
+		{
+			if (flashPaletteEffect.Lightning && info.SoundFiles.Any())
```
You can cache the `.Any()` result in the ctor, if we actually gain anything from that.

ScottNZ replied:
This is probably fine, Any() is pretty cheap to call

Mailaender replied:
Reverted it in https://github.com/OpenRA/OpenRA/pull/7863 as well.

Phrohdoh commented on OpenRA.Mods.Common/Traits/PaletteEffects/FlashPaletteEffect.cs:
```diff
@@ -0,0 +1,61 @@
+				for (var x = 0; x < 256; x++)
+				{
+					if (Lightning)
+						palette.SetColor(x, Color.White);
```
I know that this is our only use right now, but could you expose this to yaml as well?

Mailaender replied:
Okay.

Mailaender commented:
> Shouldn't this subclass the global modifier?
> Because you are just duplicating some logic.
> Especially the info property, but perhaps that is not best.

Okay, will try that. Then this depends on https://github.com/OpenRA/OpenRA/pull/7863 now.

pchote commented:
I don't like being the party pooper all the time, but I think that this use case (thunder/lighting) would be much better served by exposing a Lua API to #7863.  That would then allow mappers to do other things such as day/night cycles, turning the weather on/off, or 70's disco mode.

Mailaender commented:
When I saw how day/night-cycling is done in Tiberian Sun with triggers that also came to my mind. Traits are a bit simpler to use. We could have both ways to set them up.

pchote commented:
Traits are simple to use, but very limited.  This trait is only really useable for the one or two maps that might want to feature a thunderstorm as a gimmick.  The TS ion storm works rather differently.

Mailaender commented:
I merged https://github.com/OpenRA/OpenRA/pull/7871 and https://github.com/OpenRA/OpenRA/pull/7863 in here as this seems to be the last one of them getting merged though not for technical reasons (GlobalPaletteModifier does not share enough logic to justify sub-classing), but to add a test case (Fort Lonestar) here. All those traits work together to create the atmosphere.

Mailaender commented:
Also tried to please the reviewers a bit.

penev92 commented:
I think the lightnings need to be nerfed. Complete white is too much.

abcdefg30 commented:
Also needs a rebase.

Mailaender commented:
Can you review and merge #7871 first?

Mailaender committed:
stop all sounds on map change

Mailaender committed:
fix a crash in AmbientSound

Mailaender committed:
add sound file playing to the Lua API

Mailaender commented:
Rebased and reworked the implementation based on the negative feedback. The `Thunder` trait has been removed. Everything is triggered via the Lua API now.

Mailaender committed:
radar LightPaletteRotator is actually RA95 exclusive

Mailaender commented:
I refactored this further by merging my FlashPaletteEffect with NukePaletteEffect eliminating redundancy and making the flash effect smoother by not going for 100% color replacement.

pchote commented:
This hasn't really addressed my core comment, which is that we don't need (or IMO want) a dedicated trait for flashing the screen.  By exposing a Lua API for `GlobalLightingPaletteEffect` you allow mappers to implement any kind of lighting changes they can think of (including lightning and day / night cycles)  without requiring a bunch of overly specific traits.

Mailaender commented:
I tried that at first, but it did not work at all. Our Lua API does not support float values (which is the first problem) and then you need to work with cache and reset the old values. Then you won't have the quick fade which is important here. The Lua API for `GlobalLightingPaletteEffect` is another pull request with the use-case: day/night cycle.

Mailaender commented:
I had to rebase this.

abcdefg30 commented:
It crashes for me when using the lua function more than once.
Example script:
```lua
	Palette.FlashEffect(25)

	Trigger.AfterDelay(DateTime.Seconds(2), function()
		Palette.FlashEffect(25)
	end)
```

Gives me:
```Exception
Red Alert Mod at Version {DEV_VERSION}
Operating System: Windows (Microsoft Windows NT 6.2.9200.0)
Runtime Version: .NET CLR 4.0.30319.34209
Exception of type `System.ArgumentException`: A value of -1 is invalid for red. Red must be greater than or equal to 0 and less than or equal to 255.
(Der Wert -1 ist für red ungültig. red muss größer als oder gleich 0 und kleiner als oder gleich 255 sein.)
   at System.Drawing.Color.CheckByte(Int32 value, String name)
   at System.Drawing.Color.FromArgb(Int32 alpha, Int32 red, Int32 green, Int32 blue)
   at OpenRA.Graphics.Util.PremultipliedColorLerp(Single t, Color c1, Color c2) in \OpenRA\OpenRA.Game\Graphics\Util.cs:Line 128.
   at OpenRA.Mods.Common.Traits.FlashPaletteEffect.AdjustPalette(IReadOnlyDictionary`2 palettes)
   at OpenRA.Graphics.HardwarePalette.ApplyModifiers(IEnumerable`1 paletteMods) in \OpenRA\OpenRA.Game\Graphics\HardwarePalette.cs:Line 114.
   at OpenRA.Graphics.WorldRenderer.RefreshPalette() in \OpenRA\OpenRA.Game\Graphics\WorldRenderer.cs:Line 228.
   at OpenRA.Graphics.WorldRenderer.Draw() in \OpenRA\OpenRA.Game\Graphics\WorldRenderer.cs:Line 129.
   at OpenRA.Game.RenderTick() in \OpenRA\OpenRA.Game\Game.cs:Line 497.
   at OpenRA.Game.Loop() in \OpenRA\OpenRA.Game\Game.cs:Line 619.
   at OpenRA.Game.Run() in \OpenRA\OpenRA.Game\Game.cs:Line 637.
   at OpenRA.Program.Run(String[] args) in \OpenRA\OpenRA.Game\Support\Program.cs:Line 111.
   at OpenRA.Program.Main(String[] args) in \OpenRA\OpenRA.Game\Support\Program.cs:Line 39.

```

Mailaender commented:
Rebased and added a safeguard against that crash.

GraionDilach commented on OpenRA.Mods.RA/Traits/PaletteEffects/LightPaletteRotator.cs:
```diff
@@ -13,7 +13,7 @@
 using OpenRA.Traits;
 
-namespace OpenRA.Mods.Common.Traits
+namespace OpenRA.Mods.RA.Traits
```
Why did you changed this?

Mailaender replied:
The radar blinking palette rotation is RA95 only.

pchote replied:
Are you sure that it shouldn't be used in TD?

Mailaender replied:
That would be https://github.com/OpenRA/OpenRA/wiki/Traits#waterpaletterotation

pchote replied:
I'm talking about the blinking lights on the mammoth tank.

Mailaender replied:
The trait is not used in the `cnc` mod. Never knew about lights on the Mammoth tank.

abcdefg30 commented:
"Battlecontrol terminated" is now cut when leaving a map.

I'm also still able to reproduce https://github.com/OpenRA/OpenRA/pull/7872#issuecomment-109759334.

Mailaender commented:
I may need your full example script, because I am not able to reproduce with my latest changes.

Mailaender commented:
> "Battlecontrol terminated" is now cut when leaving a map.

Fixed by putting the stop sound into world dispose like https://github.com/OpenRA/OpenRA/pull/8568.

pchote commented on OpenRA.Game/World.cs:
```diff
@@ -387,6 +387,7 @@ public void Dispose()
 
 			frameEndActions.Clear();
 
+			Sound.StopAudio();
```
Is this still valid after all the recent churn in the sound code?

GraionDilach replied:
I don't think we have local controls/wrappers over sounds, so that shall stay, yes.

obrakmann replied:
I just noticed that this cuts off the "battlecontrol terminated" sound. It's now just "battlecontrol terminat..."

obrakmann replied:
The fix for that is easy. Just increase the `exitDelay` in `IngameMenuLogic` and `LeaveMapWidget` to 1400 milisecs.

Mailaender replied:
Updated.

pchote commented on OpenRA.Mods.Common/Scripting/Global/PaletteGlobal.cs:
```diff
@@ -0,0 +1,33 @@
+	[ScriptGlobal("Palette")]
+	public class PaletteGlobal : ScriptGlobal
+	{
+		FlashPaletteEffect fpe;
```
`readonly` please

pchote commented on mods/ra/maps/fort-lonestar/map.yaml:
```diff
@@ -486,6 +486,14 @@ Rules:
+			Blue: 1.5
+			Ambient: 0.35
+		AmbientSound:
+			SoundFile: rain.aud
```
I don't hear any rain when playing the map, only thunder.

Mailaender replied:
It is subtle, but there.

pchote commented:
The overall implementation looks fine, but we'll need to resolve the audio situation (which i've completely lost track of) before merging.

pchote commented on OpenRA.Mods.Common/Traits/PaletteEffects/FlashPaletteEffect.cs:
```diff
@@ -0,0 +1,83 @@
+			if (remainingFrames == 0 || flashing)
+				return;
+
+			flashing = true;
```
What is the purpose of this?  It looks like a race-guard, but this is single-threaded code.

Mailaender replied:
Overlapping effects such as those triggered with Lua by @abcdefg30 can lead to problems.

pchote replied:
I suggest you take a closer look at this code.  As it is written, this doesn't do anything.

pchote commented on OpenRA.Mods.Common/Scripting/Global/PaletteGlobal.cs:
```diff
@@ -0,0 +1,33 @@
+		public PaletteGlobal(ScriptContext context)
+			: base(context)
+		{
+			fpe = context.World.WorldActor.TraitOrDefault<FlashPaletteEffect>();
```
This won't scale to multiple flash effects (nuke vs lightning, for example).

pchote replied:
You could add a type argument to FlashEffect / Type field to FlashPaletteEffectInfo, cache an array of all traits, and then filter for the requested trait in FlashEffect.

Mailaender replied:
That sounds quite complex for a problem that doesn't exist yet. I would like to postpone that.

pchote replied:
Fixing this in the future would break backwards compatibility, so that would be a bad idea.

Mailaender replied:
Well you can't have nukes in Fort Lonestar, but I think your idea prevents race conditions far better than my boolean so I gave it a try.

Mailaender commented:
Rebased and applied the requested changes.

abcdefg30 commented:
Needs a rebase again.

Mailaender commented:
Rebased and updated.

obrakmann commented on OpenRA.Mods.Common/Scripting/Global/PaletteGlobal.cs:
```diff
@@ -0,0 +1,36 @@
+
+namespace OpenRA.Mods.Common.Scripting
+{
+	[ScriptGlobal("Palette")]
```
I'd prefer if this reflected the function, not the underlying mechanism. How about "Environment"? We could put all the other (global) ambient sound, weather etc. stuff into that global, too, then.

pchote replied:
Maybe `Effect`?  Nuke flashes aren't really environmental.

Mailaender replied:
I want to place other palette altering things here for day/night cycles etc.

obrakmann replied:
Day/night cycles were included in my "etc." :)

I'm just seeing this from a scripter's point of view. Many of those don't know and don't care that you're achieving the effect with modifying a palette, so referring to that in the API makes it less accessible IMO. The API should instead reflect what the scripter actually wants to do, not how it's done behind the scenes.

pchote commented on OpenRA.Mods.Common/Scripting/Global/PaletteGlobal.cs:
```diff
@@ -0,0 +1,36 @@
+		}
+
+		[Desc("Controls the `FlashPaletteEffect` trait.")]
+		public void FlashEffect(int ticks = -1, string type = null)
```
This could then be renamed to `Flash`.

pchote replied:
`type` should come before `ticks`.

obrakmann commented:
Looks good overall. :+1: once we've decided on a name for the API.

Mailaender committed:
sort globals alphabetically

Mailaender committed:
add thunderstorm effects to fort lonestar

Mailaender commented:
Renamed things.

Mailaender committed:
unhardcode and increase the exit delay

obrakmann commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/01970a08428e55e16c32010f0668b5170f4735ef%5E...01970a08428e55e16c32010f0668b5170f4735ef)
